File: backend/src/GatherGrove.Application/Services/CommunicationWorkflowService.cs
using GatherGrove.Application.DTOs;
using GatherGrove.Domain.Entities;
using GatherGrove.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace GatherGrove.Application.Services;

/// <summary>
/// Service for managing automated communication workflows
/// </summary>
public class CommunicationWorkflowService : ICommunicationWorkflowService
{
    private readonly GatherGroveDbContext _context;
    private readonly ILogger<CommunicationWorkflowService> _logger;

    public CommunicationWorkflowService(
        GatherGroveDbContext context,
        ILogger<CommunicationWorkflowService> _logger)
    {
        _context = context;
        this._logger = _logger;
    }

    public async Task<WorkflowResponse> CreateWorkflowAsync(int clubId, int userId, CreateWorkflowRequest request)
    {
        _logger.LogInformation("Creating communication workflow for club {ClubId}", clubId);

        var workflow = new CommunicationWorkflow
        {
            ClubId = clubId,
            WorkflowName = request.WorkflowName,
            TriggerType = request.TriggerType,
            WorkflowSteps = request.WorkflowSteps,
            IsActive = true,
            CreatedByUserId = userId,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _context.CommunicationWorkflows.Add(workflow);
        await _context.SaveChangesAsync();

        return MapToResponse(workflow);
    }

    public async Task<WorkflowResponse> GetWorkflowAsync(int clubId, int workflowId)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        return MapToResponse(workflow);
    }

    public async Task<List<WorkflowResponse>> GetWorkflowsAsync(int clubId, bool includeInactive = false)
    {
        var query = _context.CommunicationWorkflows
            .Where(w => w.ClubId == clubId);

        if (!includeInactive)
        {
            query = query.Where(w => w.IsActive);
        }

        var workflows = await query
            .OrderByDescending(w => w.CreatedAt)
            .ToListAsync();

        return workflows.Select(w => MapToResponse(w)).ToList();
    }

    public async Task<WorkflowResponse> UpdateWorkflowAsync(int clubId, int workflowId, UpdateWorkflowRequest request)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        if (!string.IsNullOrEmpty(request.WorkflowName))
        {
            workflow.WorkflowName = request.WorkflowName;
        }

        if (request.WorkflowSteps != null)
        {
            workflow.WorkflowSteps = request.WorkflowSteps;
        }

        if (request.IsActive.HasValue)
        {
            workflow.IsActive = request.IsActive.Value;
        }

        await _context.SaveChangesAsync();

        _logger.LogInformation("Updated workflow {WorkflowId}", workflowId);

        return MapToResponse(workflow);
    }

    public async Task DeleteWorkflowAsync(int clubId, int workflowId)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        _context.CommunicationWorkflows.Remove(workflow);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Deleted workflow {WorkflowId}", workflowId);
    }

    public async Task<WorkflowResponse> ToggleWorkflowAsync(int clubId, int workflowId, bool isActive)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        workflow.IsActive = isActive;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Toggled workflow {WorkflowId} to {IsActive}", workflowId, isActive);

        return MapToResponse(workflow);
    }

    public async Task<WorkflowStatsResponse> GetWorkflowStatsAsync(int clubId, int workflowId)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        // TODO: Implement actual stats tracking
        return new WorkflowStatsResponse
        {
            WorkflowId = workflowId,
            WorkflowName = workflow.WorkflowName,
            TotalExecutions = 0,
            SuccessfulExecutions = 0,
            FailedExecutions = 0,
            SuccessRate = 0
        };
    }

    public async Task ExecuteWorkflowAsync(int clubId, int workflowId, ExecuteWorkflowRequest request)
    {
        var workflow = await _context.CommunicationWorkflows
            .FirstOrDefaultAsync(w => w.Id == workflowId && w.ClubId == clubId);

        if (workflow == null)
        {
            throw new ArgumentException("Workflow not found");
        }

        if (!workflow.IsActive)
        {
            throw new InvalidOperationException("Workflow is not active");
        }

        _logger.LogInformation("Executing workflow {WorkflowId} for club {ClubId}", workflowId, clubId);

        // TODO: Implement workflow execution logic
        await Task.CompletedTask;
    }

    private WorkflowResponse MapToResponse(CommunicationWorkflow workflow)
    {
        return new WorkflowResponse
        {
            Id = workflow.Id,
            ClubId = workflow.ClubId,
            WorkflowName = workflow.WorkflowName,
            TriggerType = workflow.TriggerType,
            WorkflowSteps = workflow.WorkflowSteps,
            IsActive = workflow.IsActive,
            CreatedAt = workflow.CreatedAt,
            UpdatedAt = workflow.CreatedAt
        };
    }
}

File: client/src/services/communicationWorkflowService.ts
import axios from 'axios';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8050';

export interface WorkflowResponse {
  id: number;
  clubId: number;
  workflowName: string;
  description?: string;
  triggerType: string;
  workflowSteps: string;
  isActive: boolean;
  createdByUserId: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateWorkflowRequest {
  workflowName: string;
  description?: string;
  triggerType: string;
  workflowSteps: string;
}

export interface UpdateWorkflowRequest {
  workflowName?: string;
  description?: string;
  triggerType?: string;
  workflowSteps?: string;
  isActive?: boolean;
}

export interface WorkflowStatsResponse {
  workflowId: number;
  totalExecutions: number;
  successfulExecutions: number;
  failedExecutions: number;
  lastExecutedAt?: string;
  averageExecutionTime: number;
}

export interface ExecuteWorkflowRequest {
  memberId?: number;
  segmentId?: number;
}

class CommunicationWorkflowService {
  private getAuthHeaders() {
    return {
      'Content-Type': 'application/json',
      credentials: 'include' as const,
    };
  }

  async getWorkflows(clubId: number, includeInactive = false): Promise<WorkflowResponse[]> {
    const response = await axios.get<WorkflowResponse[]>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows`,
      {
        ...this.getAuthHeaders(),
        withCredentials: true,
        params: { includeInactive },
      }
    );
    return response.data;
  }

  async getWorkflow(clubId: number, workflowId: number): Promise<WorkflowResponse> {
    const response = await axios.get<WorkflowResponse>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}`,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
    return response.data;
  }

  async createWorkflow(clubId: number, request: CreateWorkflowRequest): Promise<WorkflowResponse> {
    const response = await axios.post<WorkflowResponse>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows`,
      request,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
    return response.data;
  }

  async updateWorkflow(
    clubId: number,
    workflowId: number,
    request: UpdateWorkflowRequest
  ): Promise<WorkflowResponse> {
    const response = await axios.put<WorkflowResponse>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}`,
      request,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
    return response.data;
  }

  async deleteWorkflow(clubId: number, workflowId: number): Promise<void> {
    await axios.delete(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}`,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
  }

  async toggleWorkflow(clubId: number, workflowId: number, isActive: boolean): Promise<WorkflowResponse> {
    const response = await axios.post<WorkflowResponse>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}/toggle`,
      isActive,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
    return response.data;
  }

  async getWorkflowStats(clubId: number, workflowId: number): Promise<WorkflowStatsResponse> {
    const response = await axios.get<WorkflowStatsResponse>(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}/stats`,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
    return response.data;
  }

  async executeWorkflow(
    clubId: number,
    workflowId: number,
    request: ExecuteWorkflowRequest
  ): Promise<void> {
    await axios.post(
      `${API_BASE_URL}/api/v1/clubs/${clubId}/communication-workflows/${workflowId}/execute`,
      request,
      { ...this.getAuthHeaders(), withCredentials: true }
    );
  }
}

export const communicationWorkflowService = new CommunicationWorkflowService();

